### PR TITLE
Correct the SM2 handling of DIGEST and DIGEST_SIZE parameters

### LINKED ISSUE #14873: Handling of DIGEST and DIGEST_SIZE parameters in set_ctx_params in sm2sig.c is wrong
The values are not checked in any way and they are just set in the context.

Then the mdname is just unused by the rest of the code and the size is used for some checking but if you set the value to some bogus number it will likely fail some further calls.

There is also remaining a TODO 3.0 which is not making much sense given the values are always settable.

```
static const OSSL_PARAM *sm2sig_settable_ctx_params(ossl_unused void *vpsm2ctx,
                                                    ossl_unused void *provctx)
{
    /*
     * TODO(3.0): Should this function return a different set of settable ctx
     * params if the ctx is being used for a DigestSign/DigestVerify? In that
     * case it is not allowed to set the digest size/digest name because the
     * digest is explicitly set as part of the init.
     */
    return known_settable_ctx_params;
}
```

### PR DESCRIPTION
**Pull request by t8m**
Only SM3 digest should be allowed with SM2.

Fixes #14873

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
I'm not sure about this one.  @InfoHunter, is SM2 ever used with a different digest?  Will it ever be?

**Comment by t8m** [comment]
> I'm not sure about this one. @InfoHunter, is SM2 ever used with a different digest? Will it ever be?

The current code will not handle arbitrary digest correctly. So either this is merged or there needs to be a different PR that fixes handling of arbitrary digests with SM2.

File: providers/implementations/signature/sm2sig.c [hunk 1]
```diff
@@ -108,8 +102,8 @@ static void *sm2sig_newctx(void *provctx, const char *propq)
         ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
         return NULL;
     }
-    /* don't allow to change MD, and in fact there is no such need */
-    ctx->flag_allow_md = 0;
+    ctx->mdsize = SM3_DIGEST_LENGTH;
+    strcpy(ctx->mdname, OSSL_DIGEST_NAME_SM3);
```
Comment (levitte) [inline providers/implementations/signature/sm2sig.c:106]: Why do you do this?  You use `OSSL_DIGEST_NAME_SM3` directly in the fetch anyway...

**Review by levitte — COMMENTED** [review]
I'm concerned about all the hard-coding in sm2sig.c.  How do we know that there will not be any acceptable digest other than SM3?

I'm find with having a check in one spot that sets the current limits, but would rather see the major parts of the code be flexible enough to leave room for other digests, so we don't have to revert this change in the future.

**Comment by t8m** [comment]
> I'm concerned about all the hard-coding in sm2sig.c. How do we know that there will not be any acceptable digest other than SM3?
> 
> I'm find with having a check in one spot that sets the current limits, but would rather see the major parts of the code be flexible enough to leave room for other digests, so we don't have to revert this change in the future.

OK, fixup pushed that implements this. There is now just one place where the name is checked.

File: providers/implementations/signature/sm2sig.c (hunk 1 above)
Comment (t8m) [reply providers/implementations/signature/sm2sig.c:121]: The code has now changed to use ctx->mdname so it is not hardcoded anymore. We of course still keep the default being SM3.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews!

**Pull request closed without merging**